src/math/exp_fcn_hifi.c

```diff
@@ -278,4 +284,130 @@ int32_t sofm_exp_int32(int32_t x)
 
 	return AE_MOVAD32_L(AE_MOVINT32X2_FROMINT64(ts));
 }
+
+/* Xtensa macros */
+/* Convert a float number to fractional Qnx.ny format. Note that there is no
+ * check for nx+ny number of bits to fit the word length of int. The parameter
+ * qy must be 31 or less.
+ */
+static int XT_Q_CONVERT_FLOAT(float a, int b)
+{
+	int val;
+	float x, y;
+
+	val = AE_SLAA32(1, b);
+	x = XT_FLOAT_S(val, 0);
+	y = XT_MUL_S(a, x);
+
+	return XT_TRUNC_S(y, 0);
+}
+
+/* Fractional multiplication with shift and round
+ * Note that the parameters px and py must be cast to (int64_t) if other type.
+ */
+static int XT_Q_MULTSR_32X32(int a, int b, int c, int d, int e)
+{
+	ae_int64 res;
+	int xt_o;
+	int shift;
+
+	res = AE_MUL32_LL(a, b);
+	shift = XT_SUB(XT_ADD(c, d), XT_ADD(e, 1));
+	res = AE_SRAA64(res, shift);
+	res = AE_ADD64(res, 1);
+	res = AE_SRAI64(res, 1);
+	xt_o = AE_MOVINT32_FROMINT64(res);
+
+	return xt_o;
+}
+
+/* A macro for Q-shifts */
+static int XT_Q_SHIFT_RND(int a, int b, int c)
+{
+	ae_int32 res;
+	int shift;
+
+	shift = XT_SUB(b, XT_ADD(c, 1));
+	res = AE_SRAA32(a, shift);
+	res = AE_ADD32(res, 1);
+	res = AE_SRAI32(res, 1);
+
+	return res;
+}
+
+/* Alternative version since compiler does not allow (x >> -1) */
+static int XT_Q_SHIFT_LEFT(int a, int b, int c)
+{
+	ae_int32 xt_o;
+	int shift;
+
+	shift = XT_SUB(c, b);
+	xt_o = AE_SLAA32(a, shift);
+
+	return xt_o;
+}
+
+#define ONE_Q20         XT_Q_CONVERT_FLOAT(1.0, 20)	  /* Use Q12.20 */
+#define TWO_Q27         XT_Q_CONVERT_FLOAT(2.0, 27)	  /* Use Q5.27 */
+#define MINUS_TWO_Q27   XT_Q_CONVERT_FLOAT(-2.0, 27)	  /* Use Q5.27 */
+
+/* Fixed point exponent function for approximate range -11.5 .. 7.6
+ * that corresponds to decibels range -100 .. +66 dB.
+ *
+ * The functions uses rule exp(x) = exp(x/2) * exp(x/2) to reduce
+ * the input argument for private small value exp() function that is
+ * accurate with input range -2.0 .. +2.0. The number of possible
+ * divisions by 2 is computed into variable n. The returned value is
+ * exp()^(2^n).
+ *
+ * Input  is Q5.27, -16.0 .. +16.0, but note the input range limitation
+ * Output is Q12.20, 0.0 .. +2048.0
+ */
+int32_t sofm_exp_fixed(int32_t x)
+{
+	int32_t xs;
+	int32_t y;
+	int32_t y0;
+	int i;
+	int n = 0;
+
+	if (x < XT_Q_CONVERT_FLOAT(-11.5, 27))
```

Comment:
here too

Reply:
The updated macro is now in use here.

Reply:
curious about the cycles that XT_Q_CONVERT_FLOAT takes.

Reply:
https://github.com/thesofproject/sof/pull/8435#discussion_r1404095250